swarmrl/observables/subdivided_vision_cones.py

```diff
@@ -0,0 +1,221 @@
+"""
+Computes vision cone(s).
+"""
+from typing import List
+
+import jax.numpy as jnp
+import numpy as np
+from jax import vmap
+
+from swarmrl.models.interaction_model import Colloid
+from swarmrl.observables.observable import Observable
+from swarmrl.utils.utils import calc_signed_angle_between_directors
+
+
+class SubdividedVisionCones(Observable):
+    """
+    The vision cone acts like a camera for the particles. It can either output all
+    colloids within its view, a function of the distances between one colloid and all
+    other colloids, or the normalised distance to the source.
+    """
+
+    def __init__(
+        self,
+        vision_range: float,
+        vision_half_angle: float,
+        n_cones: int,
+        radii: List[float],
+        detected_types=None,
+    ):
+        """
+        Constructor for the observable.
+
+        Parameters
+        ----------
+        vision_range : float
+                How far the particles can see.
+        vision_half_angle : float
+                the half width of the field of view in radiant units
+        n_cones: int
+                In how many cone is the field of view subdivided
+        radii: list
+                List of the radii of the colloids in the experiment,
+                ordered by the ids of the colloids
+        detected_types: list
+                list of colloid types to be detected in requested order.
+                For example [0,2] here colloids of type 1 won't be detected.
+                if None then al will be detected.
+
+        """
+        self.vision_range = vision_range
+        self.vision_half_angle = vision_half_angle
+        self.n_cones = n_cones
+        self.radii = radii
+        self.detected_types = detected_types
+
```

Comment:
you can add a jit compiled version of the angle function in the init:

```python
self.angle_fn = jax.jit(calc_signed_angle_between_directors)
```

later in the code you can then call:

```python
self.angle_fn(my_director, other_director)
```